cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/remote/RemoteLogAppender.java

```diff
@@ -0,0 +1,169 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.appender.remote;
+
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.http.DefaultHttpRequestConfig;
+import co.cask.cdap.common.internal.remote.RemoteClient;
+import co.cask.cdap.common.service.RetryStrategies;
+import co.cask.cdap.logging.appender.AbstractLogPublisher;
+import co.cask.cdap.logging.appender.LogAppender;
+import co.cask.cdap.logging.appender.LogMessage;
+import co.cask.cdap.logging.appender.kafka.LogPartitionType;
+import co.cask.cdap.logging.serialize.LoggingEventSerializer;
+import co.cask.common.http.HttpMethod;
+import co.cask.common.http.HttpRequest;
+import co.cask.common.http.HttpResponse;
+import com.google.common.hash.Hashing;
+import com.google.common.net.HttpHeaders;
+import com.google.inject.Inject;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.io.DatumWriter;
+import org.apache.avro.io.Encoder;
+import org.apache.avro.io.EncoderFactory;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.nio.ByteBuffer;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Remote log appender to push logs to log saver.
+ */
+public class RemoteLogAppender extends LogAppender {
+  private static final String APPENDER_NAME = "RemoteLogAppender";
+  private final RemoteLogPublisher publisher;
+
+  @Inject
+  public RemoteLogAppender(CConfiguration cConf, DiscoveryServiceClient discoveryServiceClient) {
+    setName(APPENDER_NAME);
+    this.publisher = new RemoteLogPublisher(cConf, discoveryServiceClient);
+  }
+
+  @Override
+  public void start() {
+    publisher.startAndWait();
+    addInfo("Successfully started " + APPENDER_NAME);
+    super.start();
+  }
+
+  @Override
+  public void stop() {
+    publisher.stopAndWait();
+    addInfo("Successfully stopped " + APPENDER_NAME);
+    super.stop();
+  }
+
+  @Override
+  protected void appendEvent(LogMessage logMessage) {
+    logMessage.prepareForDeferredProcessing();
+    logMessage.getCallerData();
+
+    try {
+      publisher.addMessage(logMessage);
+    } catch (InterruptedException e) {
+      addInfo("Interrupted when adding log message to queue: " + logMessage.getFormattedMessage());
+    }
+  }
+
+  /**
+   * Publisher service to publish logs to log saver.
+   */
+  private final class RemoteLogPublisher extends AbstractLogPublisher<Map.Entry<Integer, ByteBuffer>> {
```

Comment:
Does this class needs to be thread safe? There could be more than one threads sharing the same LogAppender?

Reply:
Good point, it should be thread safe.